Add button type to updatemenus

rreusser opened the pull request:
This PR adds the following attributes to updatemenus:

- `type`: `'buttons'`/`'dropdown'`: if true, is buttons instead of dropdown
- `direction`: `'left'`/`'right'`/`'up'`/`'down'`: affects which way the buttons extend, though they're still ordered l-to-r or top-to-bottom. It's not inconceivable to auto-detect this, but I think sometimes less magic is better.
- `showactive`: boolean. Shows active state if true. Set to false to turn the buttons into plain non-stateful buttons.

See live example at: http://rickyreusser.com/animation-experiments/#updatemenus

Notably, this PR does *not* implement scrolling for overflow. 😞

etpinard commented on src/components/updatemenus/attributes.js:
```diff
@@ -57,6 +57,37 @@ module.exports = {
+        dflt: false,
+        description: [
+            'For dropdown menus, opens the menu in the reverse direction'
+        ]
```
:eyeglasses: missing `.join(' ')`

rreusser replied:
Ah, thanks! Yes, the tests seem to check for that :)

etpinard commented on src/components/updatemenus/attributes.js:
```diff
@@ -57,6 +57,37 @@ module.exports = {
+        description: [
+            'Determines whether the menu and buttons are laid out vertically',
+            'or horizontally'
+        ]
```
another missing `.join(' ')`

etpinard commented on src/components/updatemenus/attributes.js:
```diff
@@ -57,6 +57,37 @@ module.exports = {
         ].join(' ')
     },
 
+    openreverse: {
```
I'm not fully in :heart: with this attribute name, but I can't think of anything else.

Maybe @cldougl could help.

with `openreverse: false` buttons drop down (or to the right)

![image](https://cloud.githubusercontent.com/assets/6675409/18791629/8030096a-8181-11e6-9e06-31854cd5b13d.png)


with `openreverse: true` the buttons drop up (or to the left)

![image](https://cloud.githubusercontent.com/assets/6675409/18791640/8a9f5d2e-8181-11e6-9e16-45b44f0a6153.png)

cldougl replied:
Does this have to be a boolean? My initial thought is that it may be more clear if it's called something with `direction` or `dropdowndirection`/`expanddirection` then have options `up` / `down`

For me, the I think the `reverse` is a tiny bit confusing because I think of the options being in a reversed order rather than the dropdown opening up vs down.

rreusser replied:
Default is to the right, if horizontal, or down, if vertical. Hence 'reverse' instead of just `up`/`down`.

rreusser replied:
Ahh… maybe orientation should encode both? The problem is that we have *two* orientations but *four* directions. What about `orientation: 'l'|'r'|'u'|'d'`

etpinard replied:
I like the `'l'|'r'|'u'|'d'` set of options.

But to avoid confusion, I wouldn't call the attribute `orientation` (as `orientation` is already used for legends, bar ...).

Maybe `direction: 'l'|'r'|'u'|'d'` ?

cldougl replied:
ah I see! Yeah, in that case it is much easier to keep it true/false.

cldougl replied:
+1 for `direction`
I like the explicitness of `'l'|'r'|'u'|'d'` over `true|false`

etpinard replied:
Actually, we spell `left` and `right` when specifying text anchors, so let's go with:

direction: 'left'|'right'|'u['|'down'

rreusser replied:
Changed to `direction: 'left'|'right'|'up'|'down'`. Waiting for tests to pass.

etpinard commented on src/components/updatemenus/draw.js:
```diff
@@ -114,12 +114,18 @@ module.exports = function draw(gd) {
+                drawButtons(gd, gHeader, gButton, menuOpts);
+            }
+        } else {
+            drawButtons(gd, gHeader, null, menuOpts);
```
Very lovely way to reuse the existing logic. Thanks!

I hope my code wasn't too hard to understand.

rreusser replied:
Nope! It was pretty straightforward. Was kinda able to shoehorn my needs into it, as you've seen.

etpinard commented on src/components/updatemenus/draw.js:
```diff
@@ -181,8 +191,8 @@ function drawHeader(gd, gHeader, gButton, menuOpts) {
         .text('▼');
```
Maybe we could add some logic for that `▼` character around `orientation` and `openreverse`.

For example, `orientation: 'h'` / `openreverse: true` would put some :arrow_backward: character to the left of the header text. `orientation: 'v'` / `openreverse: ture` would put some :arrow_up: character etc.

rreusser replied:
Thought about that. I had mixed feelings. I'm not opposed to it. For comparison, system select menus look like this:

<img width="229" alt="2016-09-23 11 53 55" src="https://cloud.githubusercontent.com/assets/572717/18792413/6daba5da-8184-11e6-82bd-a03a1921c957.png">

etpinard replied:
I have no strong opinion about it.

Let's ask @jackparmer and @delekru

etpinard commented on src/components/updatemenus/draw.js:
```diff
@@ -313,7 +363,7 @@ function styleButtons(buttons, menuOpts) {
         var button = d3.select(this);
 
-        if(i === active) {
+        if(i === active && menuOpts.showactive) {
```
very nice

etpinard commented on src/components/updatemenus/draw.js:
```diff
@@ -362,11 +417,51 @@ function findDimenstions(gd, menuOpts) {
+
+        // Store per-item sizes since a row of horizontal buttons, for example,
+        // don't all need to be the same width:
+        menuOpts.widths[i] = wEff;
```
Very nice. I think now this exact same `findDimensions` routine could be use for `RangeSelectors` - that will be for a future PR though.

rreusser replied:
Interesting. Yeah, I had to do some carefully wacky stuff to get it to handle all the cases (h/v, buttons/dropdown, reverse/nonreverse, anchor=t/b/l/r—which is at least no less than like 2^5 = 32 cases), but I think it works…

rreusser commented:
Fixed image mock (floating point pixel discrepancy; modified for alignment purposes). And added a pretty deluxe new one. Doesn't test *opened* menus, but I like it anyway. At least sets up difficult-to-mock extras for easy verification in dashboard.

![updatemenus_positioning](https://cloud.githubusercontent.com/assets/572717/18795757/11216b1e-8194-11e6-9aba-392d6561f1ad.png)

rreusser commented:
@etpinard the majority of the code went toward positioning so that the image mock tests the majority of what's implemented here. I could add more tests, but I think it'd only be redundant. Other than that though, I think this is ready for review.

rreusser commented on src/components/updatemenus/draw.js:
```diff
@@ -43,12 +43,12 @@ module.exports = function draw(gd) {
      *         <g item header />
      *         <text item header-arrow />
      *     <g header-group />
-     *         <g item header />
```
AHHHH stupid vim plugin. It mangles html tags when you scroll incorrectly… restoring these.

etpinard commented:
:dancer: This looks amazing 

@rreusser can you update the introduction comment block for future reference to this PR (it's still showing `openreverse` and `orientation`). Thanks!

As a side note, I'm kind of kicking myself for not naming the update menu `buttons` attribute `items` instead, having `{ type: 'buttons', buttons: [{ }, /* */] }` looks a little weird. But, I guess it's not the end of the world.

etpinard commented:
Oh and by the way, I added https://github.com/plotly/plotly.js/pull/974#discussion_r80271589 to https://github.com/plotly/plotly.js/issues/810 - if ever we decide to implement it in the future.

rreusser commented:
@etpinard yeah, backwards compatibility can be a burden! It's always possible to move to a new syntax and very slowly deprecate the old, but I think this is fine for the foreseeable future.
